### ISSUE TITLE
shellcheck treats readonly variables as non-constant

### ISSUE THREAD

**Issue by marcone** [koalaman/shellcheck#1944]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC1090
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

readonly setup_file=/path/to/setup.conf

if [ -e $setup_file ]
then
  source $setup_file
else
  echo does not exit
fi
  
```

#### Here's what shellcheck currently says:
```
Line 7	SC1090: Can't follow non-constant source. Use a directive to specify location.
```
#### Here's what I wanted or expected to see:

since "setup_file" is a read-only variable, `source $setup_file` should be the treated the same as `source /path/to/setup.conf`.

**Comment by gmolveau** [comment]
any updates about this issue ? :)

**Comment by koalaman** [comment]
@gmolveau  This is one of many issues that are currently difficult to implement robustly due to ShellCheck's lack of robust dataflow analysis, so unfortunately it'll be a while

**Comment by LLyaudet** [comment]
Hello :),

I have a similar need that doesn't use readonly keyword, but I should add it.
https://github.com/LLyaudet/DevOrSysAdminScripts/blob/7f6c55a/build_and_checks_dependencies/listings/update_or_check_files_names_listing.sh
On lines 33 and 35, and the line 37 was not reported : https://github.com/koalaman/shellcheck/issues/2986
I created this issue for the fact that rule SC1090 is not always raised.
But I have a suggestion @koalaman :

- cut the Gordian Knot : https://en.wikipedia.org/wiki/Gordian_Knot https://fr.wikipedia.org/wiki/N%C5%93ud_gordien
- You will never be able to check everything in the dataflow... because of trap built-in:
-  - `az="az"`
-  - `trap "az='za'" DEBUG`
- So, just assume that readonly is readonly and execute some kind of str_replace before processing some kinds of checks.

Best regards,
    Laurent Lyaudet
